Added Python variables example and use parameters

pvhynes opened the pull request:
- Added example which displays the variables associated with a dataset
 - Use command line arguments for query parameters
 - Added README.md in python sub-directory

pvhynes committed:
Added Python variables example and use parameters

 - Added example which displays the variables associated with a dataset
 - Use command line arguments for query parameters
 - Added README.md in python sub-directory

duncanharris commented on python/README.md:
```diff
@@ -0,0 +1,57 @@
+
+The API is powered by [GraphQL](https://graphql.org/) which allows users to request only the data that's relevant to them.
+
+The user is presented with a [GraphiQL](https://github.com/graphql/graphiql) session when the API is accessed via a web browser. GraphiQL is the GraphQL integrated development environment (IDE).
```
If you visit the API URL with a web browser then you see [GraphiQL](https://github.com/graphql/graphiql) GraphiQL web application which is a GraphQL integrated development environment (IDE).

pvhynes replied:
Drop the repeated GraphiQL?

duncanharris replied:
Yes, typo.

duncanharris commented on python/README.md:
```diff
@@ -0,0 +1,57 @@
+
+The user is presented with a [GraphiQL](https://github.com/graphql/graphiql) session when the API is accessed via a web browser. GraphiQL is the GraphQL integrated development environment (IDE).
+
+Requests to the API via other means return reponses in GraphQL format, which is a little like JSON.
```
Responses are JSON? I think loading the GraphiQL web app is not really "a request to the API", so maybe should say "If you make a request to the API URL..."

My understanding is that requests to the API are in GraphQL query language whereas responses are in JSON.

GraphiQL itself makes requests to the GraphQL API in the normal way when you submit a query in it.

pvhynes replied:
You can write queries in the GraphQL query language and submit these to the API URL. The responses will be in JSON format.
?

pvhynes replied:
You can see help text by running any of the scripts with a `-h` flag e.g.:

```
python3 variables.py -h
```

duncanharris commented on python/README.md:
```diff
@@ -0,0 +1,57 @@
+
+### `variables.py`
+
+`variables.py` is used to identify the variables associated with a specific dataset. It constructs a GraphQL query using a user-supplied base URL and dataset name and prints the names of all the available variables. The variables are separated into **rule variables** and other variables. Rule variables are special variables that are used in rule based redaction of categories when cross tabulation ...
```
`variables.py` prints out all the variables in a dataset that can be used for queries.  The variables are separated...

duncanharris replied:
Even then I'm not totally getting away from passive voice. Try again:

`variables.py` prints out all the variables in a dataset available for queries. The program separates the output into **rule variables** and the remainder. A rule variable is one which Cantabular uses for per category publication according to disclosure control rules.

pvhynes replied:
OK

pvhynes committed:
Use GraphQL variables for query parameters

mikejamesthompson commented on python/README.md:
```diff
@@ -0,0 +1,57 @@
+
+The user is presented with a [GraphiQL](https://github.com/graphql/graphiql) session when the API is accessed via a web browser. GraphiQL is the GraphQL integrated development environment (IDE).
+
+Requests to the API via other means return reponses in GraphQL format, which is a little like JSON.
```
Requests are in GraphQL format, which is a little like JSON, but the responses are just plain old JSON.

pvhynes replied:
Noted

mikejamesthompson commented on python/README.md:
```diff
@@ -0,0 +1,57 @@
+
+The API is powered by [GraphQL](https://graphql.org/) which allows users to request only the data that's relevant to them.
+
+The user is presented with a [GraphiQL](https://github.com/graphql/graphiql) session when the API is accessed via a web browser. GraphiQL is the GraphQL integrated development environment (IDE).
```
I think GraphiQL is just one IDE rather than "the" IDE.

pvhynes replied:
OK

duncanharris commented on python/variables.py:
```diff
@@ -0,0 +1,108 @@
+    parser.add_argument("base_url",
+                        metavar="URL",
+                        type=str,
+                        help="Cantabular GraphQL base URL including port "
```
Cantabular extended API base URL?

pvhynes replied:
OK

duncanharris commented on python/README.md:
```diff
@@ -0,0 +1,57 @@
+`query.py` is used to perform cross tabulations and print the resulting table in CSV format. It constructs a GraphQL query using a user-supplied base URL, dataset name and variable list.
+
+
+The script takes at least 3 parameters. The first is the base URL for the GraphQL endpoint, whilst the second is the name of the dataset. All subsequent parameters are treated as the names of query variables. At most one rule variable can be specified and this must be the first supplied variable. This example performs a cross tabulation for a locally hosted dataset called `Example` using the vari ...
```
Should we be duplicating what is in the command line help? Should we teach the user to fish? i.e. tell them to run the script to get help? Wondering... I'm very averse to duplication: more maintenance and risk of inconsistency.

pvhynes replied:
In the "Scripts" subsection I'll mention the help option

duncanharris commented on python/query.py:
```diff
@@ -90,5 +92,34 @@ def main():
+                        nargs="+",
+                        type=str,
+                        help="Names of variables to use in query. At least one variable name "
+                        "must be supplied. A single rule variable may be specified. If a rule "
```
You can supply at most one rule variable which must be first if present.
?

pvhynes replied:
OK

mikejamesthompson reviewed (changes requested):
Left a couple of comments on the docs which should be changed.

Tested both scripts on the Ireland census API on macOS and they both run with no issues.

Do we need the licence in the individual files if we have one in the root of the repo?

duncanharris commented on python/variables.py:
```diff
@@ -0,0 +1,108 @@
+    print("")
+    print("OTHER VARIABLES")
+    print("---------------")
+    for variable in variables - rule_variables:
```
I thought `variables - rule_variables` was a bit non-obvious. Maybe we should use brackets or assign to a separate variable?, e.g.
```
non_rule_variables = variables - rule_variables
```

pvhynes replied:
OK

duncanharris commented on python/query.py:
```diff
@@ -90,5 +92,34 @@ def main():
+    parser.add_argument("base_url",
+                        metavar="URL",
+                        type=str,
+                        help="Cantabular GraphQL base URL including port "
```
Cantabular extended API base URL?

pvhynes replied:
OK

duncanharris commented on python/query.py:
```diff
@@ -51,9 +50,12 @@ def main():
 
     Query the Cantabular extended API, verify and parse the response
-    and print the rows containing in the table in CSV format.
+    and print the rows contained in the table in CSV format.
```
s/contained in/of/
?

pvhynes replied:
OK

duncanharris reviewed (changes requested):
Prefer active voice: https://developers.google.com/tech-writing/one/active-voice

pvhynes committed:
Changes following PR

 - Updated help text
 - Clearer instructions in README.md with less duplication
 - New non_rule_variables variable to make intent more obvious

mikejamesthompson reviewed (approved):
LGTM

duncanharris reviewed (approved):
LGTM
